ksqldb-engine/src/main/java/io/confluent/ksql/engine/TransientQueryCleanupService.java

```diff
@@ -0,0 +1,276 @@
+        if (allTransientQueriesEverCreatedIds.contains(leakingQueryId)
+                && isTerminatedQuery(leakingQueryId)) {
+          LOG.info("{} is leaking topics. Adding it to the cleanup queue.", leakingQueryId);
+          addCleanupTask(new TransientQueryTopicCleanupTask(
+                  serviceContext,
+                  leakingQueryId
+          ));
+        }
+      }
+    }
+  }
+
+  public void recordTransientQueryOnStart(final TransientQueryMetadata queryMetadata) {
+    this.allTransientQueriesEverCreated.add(queryMetadata);
+    this.allTransientQueriesEverCreatedIds.add(queryMetadata.getQueryApplicationId());
+  }
+
+  public static class TransientQueryTopicCleanupTask implements Callable<Boolean> {
+
+    private final ServiceContext serviceContext;
+    private final String appId;
+
+    public TransientQueryTopicCleanupTask(
+            final ServiceContext serviceContext,
+            final String appId
+    ) {
+      this.serviceContext = Objects.requireNonNull(serviceContext, "serviceContext");
+      this.appId = Objects.requireNonNull(appId, "appId");
+    }
+
+    @Override
+    public Boolean call() {
+      try {
+        LOG.info("Deleting topics for transient query: {}", appId);
+        serviceContext.getTopicClient().deleteInternalTopics(appId);
+      } catch (final Exception e) {
+        LOG.warn("Failed to cleanup topics for transient query: {}", appId, e);
+        return false;
+      }
+
+      final Set<String> orphanedTopics = serviceContext
+              .getTopicClient()
+              .listTopicNames()
+              .stream()
+              .filter(topicName -> topicName.startsWith(appId))
+              .collect(Collectors.toSet());
+
+      if (orphanedTopics.isEmpty()) {
+        return true;
+      }
+
+      LOG.warn("Failed to cleanup topics for transient query: {}\n"
+                      + "The following topics seem to have leaked: {}",
+              appId,
+              orphanedTopics);
+
+      return true;
+    }
+  }
+
+  public static class TransientQueryStateCleanupTask implements Callable<Boolean> {
+    private final String appId;
+    private final String pathName;
+
+    public TransientQueryStateCleanupTask(
+            final String appId,
+            final String stateDir
+    ) {
+      this.appId = Objects.requireNonNull(appId, "appId");
+      this.pathName = stateDir + "/" + appId;
+    }
+
+    @Override
+    public Boolean call() throws IOException {
+      final Path pathName = Paths.get(this.pathName);
+      final File directory = new File(String.valueOf(pathName.normalize()));
+      deleteIfExists(pathName);
+      if (directory.exists()) {
+        LOG.warn("Failed to delete local state store for non-existing query: {}",
+                appId);
+        return false;
+      } else {
+        return true;
+      }
+    }
+  }
+
+  private boolean isTerminatedQuery(final String queryId) {
+    return !allTransientQueriesEverCreated
```

Comment:
We can always grab the metadata from the query registry using the id to avoid the memory issues.  If it's not there, it's safe to assume that the query has been stopped.

Reply:
good idea. Changed to query the `QueryRegistry` to check the status of the query (running or not) instead of maintaining a set of all the queries ever created.